### PR TITLE
Refactor into MerkleTree structure

### PR DESCRIPTION
**Pull request by dsocolobsky**
This does not add new functionality just refactors the code a bit so that this PR doesn't mix with further commits.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,58 +1,98 @@
...
+    /// Checks if the given hash is present at any level of the Merkle Tree
+    fn contains_hash(&self, hash: &String) -> bool {
+        self.levels.iter().any(|level| level.contains(hash))
+    }
+
+    /// Generates a proof that a certain element belongs to the tree, if present.
+    fn generate_proof(&self, element: u32) -> Option<(Vec<String>, usize)> {
+        let Some( leaf_index) = self.leaf_index_for_element(element) else {
+            return None
         };
```
Comment (LeanSerra) [inline src/main.rs:59]: You can use the `?` operator so it looks more clean
`let leaf_index = self.leaf_index_for_element(element)?;`

File: src/main.rs [hunk 2]
```diff
@@ -1,58 +1,98 @@
...
+        let mut proof: Vec<String> = vec![];
+
+        let mut index = leaf_index;
+        for level in self.levels.iter().take(self.height() - 1) {
+            if index % 2 == 0 {
+                proof.push(level[index+1].clone())
+            } else {
+                proof.push(level[index-1].clone())
+            }
+            index = index / 2;
```
Comment (LeanSerra) [inline src/main.rs:69]: This can be rewritten as `index /= 2`

File: src/main.rs [hunk 3]
```diff
@@ -1,58 +1,98 @@
...
+        };
+        let mut hash = hash_one(element);
+        let mut index = index;
+        for p in proof.iter() {
+            hash = if index % 2 == 0 {
+                hash_multiple(vec![hash.clone(), p.clone()])
+            } else {
+                hash_multiple(vec![p.clone(), hash.clone()])
+            };
+            index = index / 2;
```
Comment (LeanSerra) [inline src/main.rs:88]: Same as above

**Review by LeanSerra — APPROVED** [review]
Some styling corrections but LGTM

**Review by damiramirez — APPROVED** [review]

File: src/main.rs (hunk 1 above)
Comment (dsocolobsky) [reply src/main.rs]: Done

**Pull request merged**